More dynamicDowncast<> adoption in accessibility code

annevk opened the pull request:
#### 9aa547bb5ff18e5ecfd244ec5ba4b7c9632e25d0
<pre>
More dynamicDowncast&lt;&gt; adoption in accessibility code
<a href="https://bugs.webkit.org/show_bug.cgi?id=269170">https://bugs.webkit.org/show_bug.cgi?id=269170</a>

Reviewed by Tyler Wilcock.

* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::computeAccessibilityIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::isTableRow const):
(WebCore::AccessibilityTableRow::parentTable const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueStroke):

Canonical link: <a href="https://commits.webkit.org/274722@main">https://commits.webkit.org/274722@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd8d3f96c0e2028703333cc8a0bbcefee5495a28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18630 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42186 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35551 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41957 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21529 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33101 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40225 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15714 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34379 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13628 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35271 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43463 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36004 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35604 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39390 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11915 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37667 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16069 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8937 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16118 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1357,7 +1352,7 @@ bool AccessibilityRenderObject::computeAccessibilityIsIgnored() const
-    if (!m_renderer || !is<RenderView>(*m_renderer))
+    if (!is<RenderView>(m_renderer))
         return 0;
     return downcast<RenderView>(*m_renderer).frameView().layoutContext().layoutCount();
```
The compiler would not let me convert this one. Something didn't line up.

annevk replied:
It was similar to what GCC apparently outputs:
```
/app/webkit/Source/WebCore/accessibility/AccessibilityRenderObject.cpp:2486:70: note:   mismatched types ‘auto*’ and ‘WTF::WeakPtr<WebCore::RenderListItem, WTF::SingleThreadWeakPtrImpl, WTF::RawPtrTraits<WTF::SingleThreadWeakPtrImpl> >’
 2486 |     auto* renderListItem = dynamicDowncast<RenderListItem>(m_renderer);
```
@cdumez ideas?

cdumez replied:
`auto* renderListItem = dynamicDowncast<RenderListItem>(m_renderer.get());` would work. `m_renderer` is a WeakPtr and `dynamicDowncast<>()` doesn't know how to handle WeakPtrs (or CheckedPtrs) at the moment. Only raw pointers and Ref/RefPtr are currently supported.

annevk commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -64,22 +64,20 @@ Ref<AccessibilityTableCell> AccessibilityTableCell::create(Node& node)
     // Ignore anonymous table cells as long as they're not in a table (ie. when display:table is used).
-    RenderObject* renderTable = is<RenderTableCell>(renderer()) ? downcast<RenderTableCell>(*m_renderer).table() : nullptr;
+    auto* renderTableCell = dynamicDowncast<RenderTableCell>(renderer());
+    RenderObject* renderTable = renderTableCell ? renderTableCell->table() : nullptr;
```
If I used auto* here the compiler would complain that node() was already deleted...

cdumez replied:
Use `auto*` and then call `element()` instead of `node()` below. `RenderElement::node()` is deleted because we know the associated Node is actually an Element and you're meant to call `RenderElement::element()` instead.

annevk commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -73,8 +73,8 @@ AccessibilityRole AccessibilityTableRow::determineAccessibilityRole()
 {
-    AccessibilityObject* table = parentTable();
-    return is<AccessibilityTable>(table) && downcast<AccessibilityTable>(*table).isExposable();
+    auto* table = parentTable();
```
parentTable() already returns AccessibilityTable*.

annevk committed:
More dynamicDowncast<> adoption in accessibility code
https://bugs.webkit.org/show_bug.cgi?id=269170

Reviewed by Tyler Wilcock.

* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::computeAccessibilityIsIgnored const):
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::isTableRow const):
(WebCore::AccessibilityTableRow::parentTable const):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueStroke):

Canonical link: https://commits.webkit.org/274722@main
